### ISSUE TITLE
Optional chaining breaking function inference.

### ISSUE THREAD

**Issue by ShanonJackson** [microsoft/TypeScript#35274]
**TypeScript Version:**  3.7.2

**Search Terms:**
optional chaining
**Code**

```ts
interface Maybe<A> { };
export declare function Maybe<A>(value: A | null | undefined): Maybe<A>;
declare const ctx: {req?: {headers?: {cookie?: string}}}
const testopt = ctx.req?.headers?.cookie
const teststring = "" as string | undefined;
const test = Maybe(testopt) // Maybe<string | undefined> (wrong)
const test1 = Maybe(teststring ) // Maybe<string>  (correct)
```

**Expected behavior:**
`const test` should be of type `Maybe<string>` like `const test1` because this is the behavior of inference on a function with this signature.

**Actual behavior:**
`const test` is of type `Maybe<string | undefined`;

**Comment by fatcerberus** [comment]
This is quite bizarre.  It would be understandable if there were some bug where it inferred different types based on the form of the expression passed to `Maybe()`.  But here they're both assigned to a variable first, which *should* equalize things--yet it doesn't.

Since the type of `cookie` is already `string | undefined`, could it be that the optional chain is actually yielding a type of `string | undefined | undefined` and this is not being properly normalized... only to become observable later when used as an inference source for `T`.

**Comment by MartinJohns** [comment]
Playground link:  
https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgLJwJ4CMIB4CCAfMgN7IC+A3ALABQEAHgA4D2UYyAJhAgDZxQUMAK4gEYYCxBpMOAoQAUANzi9hEAFzJ8yAD7IQw3rz3JR3GKAicAlFvTY8RGrW58BKBFIDOHcQy0SQQBHAH5AgAsIOG4ob3DSLxYAa2AIBN8oUABzcjy6LxBfZEhfFiYOAF5kfwA6ENDaqJjoeNqk1IgCnw5SsEyc5GqAImHkOG9kAZBs03MISxBrF0LivqGZRwU+8rAbZAB6A825adn9ecXrYgUAdygpbJtuot6IXwBGDYccbff+sBZGb7I4nPBnYjIBReKCCcTPWhAA

**Comment by ajafff** [comment]
Should be fixed by #34588 which is already available in `typescript@next` and will be in the next patch release.

**Comment by RyanCavanaugh** [comment]
Y'all know you can try "Nightly" in the Playground, right?

Duplicate #34579
